12:25:15       [exec] Caused by: java.lang.NullPointerException
12:25:15       [exec] 	at com.sun.s1peqe.security.ssl.converter.ejb.ConverterBean.myCallerPrincipal(ConverterBean.java:38)

--- a/nucleus/security/core/src/main/java/com/sun/enterprise/security/SecurityContext.java
+++ b/nucleus/security/core/src/main/java/com/sun/enterprise/security/SecurityContext.java
@@ -75,7 +75,7 @@ public class SecurityContext extends AbstractSecurityContext {
     // Did the client log in as or did the server generate the context
     private boolean serverGeneratedSecurityContext;
 
-    private Principal sessionPrincipal;
+    private ThreadLocal<Principal> sessionPrincipal = new ThreadLocal<>();
 
     /*
      * This creates a new SecurityContext object. Note: that the docs for Subject state that the internal sets (eg. the
@@ -289,8 +289,7 @@ public class SecurityContext extends AbstractSecurityContext {
     public static SecurityContext getCurrent() {
         SecurityContext securityContext = currentSecurityContext.get();
         if (securityContext == null) {
-            securityContext = generateDefaultSecurityContext();
-            currentSecurityContext.set(securityContext);
+            securityContext = defaultSecurityContext;
         }
 
         return securityContext;
@@ -371,11 +370,15 @@ public class SecurityContext extends AbstractSecurityContext {
     }
 
     public Principal getSessionPrincipal() {
-        return sessionPrincipal;
+        return sessionPrincipal.get();
     }
 
     public void setSessionPrincipal(Principal sessionPrincipal) {
-        this.sessionPrincipal = sessionPrincipal;
+        if (sessionPrincipal != null) {
+            this.sessionPrincipal.set(sessionPrincipal);
+        } else {
+            this.sessionPrincipal.remove();
+        }
     }
 
     public Set<Principal> getPrincipalSet() {

